Mathlib/Algebra/Group/Defs.lean

```diff
@@ -579,6 +579,18 @@ class AddMonoid (M : Type u) extends AddSemigroup M, AddZeroClass M where
   nsmul_succ : ∀ (n : ℕ) (x), nsmul (n + 1) x = x + nsmul n x := by intros; rfl
 #align add_monoid AddMonoid
 
+/-
+If
+structure AB extends A
+structure AC extends A
+structure ABC extends AB, AC
+then AB.toA should have higher priority than AC.toA
+Todo: write linter.
```

Comment:
Can this be done automatically? I mean, when declaring a new class, put priority 200 on classes it extends with a `toA`, priority 100 otherwise (and maybe priority 1000 if it is a specific instance that doesn't always apply, like we did in mathlib3)?

Reply:
I think it's worth opening a lean4 issue about this

Reply:
What's supposed to happen here?

```
structure AB extends A
structure AC extends A
structure ABC extends AB, AC
structure ACB extends AC. AB
```
Do we need a weight-style system like Coq, where the priority is the sum of all the paths, instead of a global priority on each step?

Reply:
I opened https://github.com/leanprover/lean4/issues/2325

@eric-wieser: just warn the user that they do something stupid, and ask them to reorder the extended structures. The sum of priorities is not good: you want to make sure that all the highest priority paths only use the "correct" projections.

Reply:
I worry that this effectively forces us to choose a global partial order over all typeclasses that exist; potentially something as silly as "the `extends` clause must be in alphabetic order"...

Reply:
`LeftCancelMonoid`, `RightCancelMonoid` and `CommMonoid` all extend `Monoid`, and if we declare commutativity to be our favourite path (it's surely the most common path) then `CancelMonoid`, which extends `LeftCancelMonoid` and `RightCancelMonoid`, is left with no "correct" projection.